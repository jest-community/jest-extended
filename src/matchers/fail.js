export function fail(_, message) {
  return {
    pass: false,
    message: () => (message ? message : 'fails by .fail() assertion'),
  };
}
