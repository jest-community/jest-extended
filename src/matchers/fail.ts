export function fail(_: unknown, message: string) {
  return {
    pass: false,
    message: () => (message ? message : 'fails by .fail() assertion'),
  };
}
