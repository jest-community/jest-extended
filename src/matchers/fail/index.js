import predicate from './predicate';

const failMessage = message => {
  if (message) return () => message;
  else return () => 'fails by .fail() assertion';
};

export default {
  fail: (expected, message) => ({ pass: predicate(), message: failMessage(message) }),
};
