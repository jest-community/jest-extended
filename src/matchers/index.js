import toBeFalse from './toBeFalse';
import toBeTrue from './toBeTrue';

export default [toBeFalse, toBeTrue].reduce((acc, matcher) => ({ ...acc, ...matcher }), {});
