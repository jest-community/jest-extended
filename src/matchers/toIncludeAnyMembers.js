import { contains } from '../utils';

export function toIncludeAnyMembers(actual, expected) {
  const { printReceived, printExpected, matcherHint } = this.utils;

  const pass =
    Array.isArray(actual) && Array.isArray(expected) && expected.some(member => contains(this.equals, actual, member));

  return {
    pass,
    message: () =>
      pass
        ? matcherHint('.not.toIncludeAnyMembers') +
          '\n\n' +
          'Expected list to not include any of the following members:\n' +
          `  ${printExpected(expected)}\n` +
          'Received:\n' +
          `  ${printReceived(actual)}`
        : matcherHint('.toIncludeAnyMembers') +
          '\n\n' +
          'Expected list to include any of the following members:\n' +
          `  ${printExpected(expected)}\n` +
          'Received:\n' +
          `  ${printReceived(actual)}`,
  };
}
