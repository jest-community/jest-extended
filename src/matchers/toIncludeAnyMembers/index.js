import { matcherHint, printExpected, printReceived } from 'jest-matcher-utils';

import predicate from './predicate';

const passMessage = (actual, expected) => () =>
  matcherHint('.not.toIncludeAnyMembers') +
  '\n\n' +
  'Expected list to not include any of the following members:\n' +
  `  ${printExpected(expected)}\n` +
  'Received:\n' +
  `  ${printReceived(actual)}`;

const failMessage = (actual, expected) => () =>
  matcherHint('.toIncludeAnyMembers') +
  '\n\n' +
  'Expected list to include any of the following members:\n' +
  `  ${printExpected(expected)}\n` +
  'Received:\n' +
  `  ${printReceived(actual)}`;

export default {
  toIncludeAnyMembers: (actual, expected) => {
    const pass = predicate(actual, expected);
    if (pass) {
      return { pass: true, message: passMessage(actual, expected) };
    }

    return { pass: false, message: failMessage(actual, expected) };
  }
};
