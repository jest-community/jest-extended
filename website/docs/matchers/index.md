---
sidebar_position: 1
---

# Matchers

- [.pass(message)](/docs/matchers/pass)
- [.fail(message)](/docs/matchers/fail)
- [.toBeEmpty()](/docs/matchers/tobeempty)
- [.toBeOneOf([members])](/docs/matchers/tobeoneof)
- [.toBeNil()](/docs/matchers/tobenil)
- [.toSatisfy(predicate)](/docs/matchers/tosatisfy)

## [Array](/docs/matchers/array)

- [.toBeArray()](/docs/matchers/array/#tobearray)
- [.toBeArrayOfSize()](/docs/matchers/array/#tobearrayofsize)
- [.toIncludeAllMembers([members])](/docs/matchers/array/#toincludeallmembersmembers)
- [.toIncludeAllPartialMembers([members])](/docs/matchers/array/#toincludeallpartialmembersmembers)
- [.toIncludeAnyMembers([members])](/docs/matchers/array/#toincludeanymembersmembers)
- [.toIncludeSameMembers([members])](/docs/matchers/array/#toincludesamemembersmembers)
- [.toPartiallyContain(member)](/docs/matchers/array/#topartiallycontainmember)
- [.toSatisfyAll(predicate)](/docs/matchers/array/#tosatisfyallpredicate)
- [.toSatisfyAny(predicate)](/docs/matchers/array/#tosatisfyanypredicate)
- [.toBeInRange(min, max)](/docs/matchers/array/#tobeinrangemin-max)

## [Boolean](/docs/matchers/boolean)

- [.toBeBoolean()](/docs/matchers/boolean/#tobeboolean)
- [.toBeTrue()](/docs/matchers/boolean/#tobetrue)
- [.toBeFalse()](/docs/matchers/boolean/#tobefalse)

## [Date](/docs/matchers/date)

- [.toBeDate()](/docs/matchers/date/#tobedate)
- [.toBeValidDate()](/docs/matchers/date/#tobevaliddate)
- [.toBeAfter(date)](/docs/matchers/date/#tobeafterdate)
- [.toBeBefore(date)](/docs/matchers/date/#tobebeforedate)
- [.toBeAfterOrEqualTo(date)](/docs/matchers/date/#tobeafterorequaltodate)
- [.toBeBeforeOrEqualTo(date)](/docs/matchers/date/#tobebeforeorequaltodate)
- [.toBeBetween(startDate, endDate)](/docs/matchers/date/#tobebetweenstartdate-enddate)

## [Function](/docs/matchers/function)

- [.toBeFunction()](/docs/matchers/function/#tobefunction)
- [.toThrowWithMessage()](/docs/matchers/function/#tothrowwithmessagetype-message)

## [Mock](/docs/matchers/mock)

- [.toHaveBeenCalledBefore()](/docs/matchers/mock/#tohavebeencalledbefore)
- [.toHaveBeenCalledAfter()](/docs/matchers/mock/#tohavebeencalledafter)
- [.toHaveBeenCalledOnce()](/docs/matchers/mock/#tohavebeencalledonce)
- [.toHaveBeenCalledOnceWith()](/docs/matchers/mock/#tohavebeencalledoncewith)

## [Number](/docs/matchers/number)

- [.toBeNumber()](/docs/matchers/number/#tobenumber)
- [.toBeNaN()](/docs/matchers/number/#tobenan)
- [.toBeFinite()](/docs/matchers/number/#tobefinite)
- [.toBePositive()](/docs/matchers/number/#tobepositive)
- [.toBeNegative()](/docs/matchers/number/#tobenegative)
- [.toBeEven()](/docs/matchers/number/#tobeeven)
- [.toBeOdd()](/docs/matchers/number/#tobeodd)
- [.toBeWithin(start, end)](/docs/matchers/number/#tobewithinstart-end)
- [.toBeInteger()](/docs/matchers/number/#tobeinteger)

## [Object](/docs/matchers/object)

- [.toBeObject()](/docs/matchers/object/#tobeobject)
- [.toBeEmptyObject()](/docs/matchers/object/#tobeemptyobject)
- [.toContainKey(key)](/docs/matchers/object/#tocontainkeykey)
- [.toContainKeys([keys])](/docs/matchers/object/#tocontainkeyskeys)
- [.toContainAllKeys([keys])](/docs/matchers/object/#tocontainallkeyskeys)
- [.toContainAnyKeys([keys])](/docs/matchers/object/#tocontainanykeyskeys)
- [.toContainValue(value)](/docs/matchers/object/#tocontainvaluevalue)
- [.toContainValues([values])](/docs/matchers/object/#tocontainvaluesvalues)
- [.toContainAllValues([values])](/docs/matchers/object/#tocontainallvaluesvalues)
- [.toContainAnyValues([values])](/docs/matchers/object/#tocontainanyvaluesvalues)
- [.toContainEntry([key, value])](/docs/matchers/object/#tocontainentrykey-value)
- [.toContainEntries([[key, value]])](/docs/matchers/object/#tocontainentrieskey-value)
- [.toContainAllEntries([[key, value]])](/docs/matchers/object/#tocontainallentrieskey-value)
- [.toContainAnyEntries([[key, value]])](/docs/matchers/object/#tocontainanyentrieskey-value)
- [.toBeExtensible()](/docs/matchers/object/#tobeextensible)
- [.toBeFrozen()](/docs/matchers/object/#tobefrozen)
- [.toBeSealed()](/docs/matchers/object/#tobesealed)

## [Promise](/docs/matchers/promise)

- [.toResolve()](/docs/matchers/promise/#toresolve)
- [.toReject()](/docs/matchers/promise/#toreject)

## [String](/docs/matchers/string)

- [.toBeString()](/docs/matchers/string/#tobestring)
- [.toBeHexadecimal(string)](/docs/matchers/string/#tobehexadecimal)
- [.toBeDateString(string)](/docs/matchers/string/#tobedatestringstring)
- [.toEqualCaseInsensitive(string)](/docs/matchers/string/#toequalcaseinsensitivestring)
- [.toStartWith(prefix)](/docs/matchers/string/#tostartwithprefix)
- [.toEndWith(suffix)](/docs/matchers/string/#toendwithsuffix)
- [.toInclude(substring)](/docs/matchers/string/#toincludesubstring)
- [.toIncludeRepeated(substring, times)](/docs/matchers/string/#toincluderepeatedsubstring-times)
- [.toIncludeMultiple([substring])](/docs/matchers/string/#toincludemultiplesubstring)
- [.toEqualIgnoringWhitespace(string)](/docs/matchers/string/#toequalignoringwhitespacestring)

## [Symbol](/docs/matchers/symbol)

- [.toBeSymbol()](/docs/matchers/symbol/#tobesymbol)
