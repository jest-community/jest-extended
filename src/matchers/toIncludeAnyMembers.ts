import { contains } from 'src/utils';

export function toIncludeAnyMembers<E = unknown>(actual: unknown, expected: readonly E[] | E) {
  // @ts-expect-error OK to have implicit any for this.utils
  const { printReceived, printExpected, matcherHint } = this.utils;

  const pass =
    // @ts-expect-error OK to have implicit any for this.equals
    Array.isArray(actual) && Array.isArray(expected) && expected.some(member => contains(this.equals, actual, member));

  return {
    pass,
    message: () =>
      pass
        ? matcherHint('.not.toIncludeAnyMembers') +
          '\n\n' +
          'Expected list to not include any of the following members:\n' +
          `  ${printExpected(expected)}\n` +
          'Received:\n' +
          `  ${printReceived(actual)}`
        : matcherHint('.toIncludeAnyMembers') +
          '\n\n' +
          'Expected list to include any of the following members:\n' +
          `  ${printExpected(expected)}\n` +
          'Received:\n' +
          `  ${printReceived(actual)}`,
  };
}
