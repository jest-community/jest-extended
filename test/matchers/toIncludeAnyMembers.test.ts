import * as matcher from 'src/matchers/toIncludeAnyMembers';

expect.extend(matcher);

describe('.toIncludeAnyMembers', () => {
  const shallow = { hello: 'world' };
  const deep = { message: shallow };

  test('passes when given array is a subset', () => {
    expect([2, 3]).toIncludeAnyMembers([1, 2, 3]);
  });

  test('passes when given array contains some of the members', () => {
    expect([1, 2, 3]).toIncludeAnyMembers([2, 3, 4]);
  });

  test('passes when given array contains object value', () => {
    expect([shallow]).toIncludeAnyMembers([shallow, deep]);
  });

  test('fails when given array does not contain any members ', () => {
    expect(() => expect([4, 5, 6]).toIncludeAnyMembers([1, 2, 3])).toThrowErrorMatchingSnapshot();
  });

  test('fails when given object is not an array', () => {
    expect(() => expect(7).toIncludeAnyMembers([1, 2, 3])).toThrowErrorMatchingSnapshot();
  });

  test('fails when expected object does not contain array value', () => {
    expect(() => expect([1, 2, 3]).toIncludeAnyMembers('world')).toThrowErrorMatchingSnapshot();
  });
});

describe('.not.toIncludeAnyMembers', () => {
  const shallow = { hello: 'world' };

  test('passes when given array does not contain primitive value', () => {
    expect(['hola', 'mundo']).not.toIncludeAnyMembers(['hello', 'world']);
  });

  test('passes when given array does not contain object value', () => {
    expect([{ hello: 'world' }]).not.toIncludeAnyMembers([{ world: 'hello' }]);
  });

  test('passes when given object is not array', () => {
    expect(7).not.toIncludeAnyMembers([7]);
  });

  test('fails when given object contains primitive value', () => {
    expect(() => expect([7]).not.toIncludeAnyMembers([7, 8])).toThrowErrorMatchingSnapshot();
  });

  test('fails when given array contains object value', () => {
    expect(() => expect([{ foo: 'bar' }]).not.toIncludeAnyMembers([{ foo: 'bar' }])).toThrowErrorMatchingSnapshot();
  });

  test('fails when given array contains array value', () => {
    expect(() => expect([[shallow]]).not.toIncludeAnyMembers([[shallow], 7])).toThrowErrorMatchingSnapshot();
  });
});

// Note - custom equality tester must be at the end of the file because once we add it, it cannot be removed
describe('toIncludeAnyMembers with custom equality tester', () => {
  let mockEqualityTester;
  beforeAll(() => {
    mockEqualityTester = jest.fn();
    expect.addEqualityTesters([mockEqualityTester]);
  });
  afterEach(() => {
    mockEqualityTester.mockReset();
  });
  test('passes when custom equality matches one of the values', () => {
    mockEqualityTester.mockImplementation((a, b) => (a === 1 && b === 4 ? true : undefined));
    expect([1]).toIncludeAnyMembers([4]);
  });
  test('fails when custom equality does not match any of the values', () => {
    mockEqualityTester.mockReturnValue(false);
    expect(() => expect([1]).toIncludeAnyMembers([1])).toThrowErrorMatchingSnapshot();
  });
});
