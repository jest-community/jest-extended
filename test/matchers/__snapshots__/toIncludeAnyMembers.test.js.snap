// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.not.toIncludeAnyMembers fails when given array contains array value 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).not.toIncludeAnyMembers(</intensity><green>expected</color><dim>)</intensity>

Expected list to not include any of the following members:
  <green>[[{\\"hello\\": \\"world\\"}], 7]</color>
Received:
  <red>[[{\\"hello\\": \\"world\\"}]]</color>"
`;

exports[`.not.toIncludeAnyMembers fails when given array contains object value 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).not.toIncludeAnyMembers(</intensity><green>expected</color><dim>)</intensity>

Expected list to not include any of the following members:
  <green>[{\\"foo\\": \\"bar\\"}]</color>
Received:
  <red>[{\\"foo\\": \\"bar\\"}]</color>"
`;

exports[`.not.toIncludeAnyMembers fails when given object contains primitive value 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).not.toIncludeAnyMembers(</intensity><green>expected</color><dim>)</intensity>

Expected list to not include any of the following members:
  <green>[7, 8]</color>
Received:
  <red>[7]</color>"
`;

exports[`.toIncludeAnyMembers fails when expected object does not contain array value 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).toIncludeAnyMembers(</intensity><green>expected</color><dim>)</intensity>

Expected list to include any of the following members:
  <green>\\"world\\"</color>
Received:
  <red>[1, 2, 3]</color>"
`;

exports[`.toIncludeAnyMembers fails when given array does not contain any members  1`] = `
"<dim>expect(</intensity><red>received</color><dim>).toIncludeAnyMembers(</intensity><green>expected</color><dim>)</intensity>

Expected list to include any of the following members:
  <green>[1, 2, 3]</color>
Received:
  <red>[4, 5, 6]</color>"
`;

exports[`.toIncludeAnyMembers fails when given object is not an array 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).toIncludeAnyMembers(</intensity><green>expected</color><dim>)</intensity>

Expected list to include any of the following members:
  <green>[1, 2, 3]</color>
Received:
  <red>7</color>"
`;
